clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -182,14 +183,20 @@ public void done(long baseOffset, long logAppendTime, RuntimeException exception
         Iterator<MutableRecordBatch> recordBatchIter = memoryRecords.batches().iterator();
         if (!recordBatchIter.hasNext())
             throw new IllegalStateException("Cannot split an empty producer batch.");
+
         RecordBatch recordBatch = recordBatchIter.next();
+        if (recordBatch.magic() < MAGIC_VALUE_V2 && !recordBatch.isCompressed())
```

Comment:
OK, I think I'm just going to add the check in `Sender.completeBatch()` and follow the behavior prior to KIP-126 if it is the old message format without compression.